Replace system readelf with pyelftools for multi-arch support

carlpinto25 opened the pull request:
This PR replaces the dependency on the system's readelf command with the pyelftools library. This addresses issue #3295 by allowing pwndbg to correctly analyze binaries from foreign architectures (e.g., ARM on x86) without relying on the host's toolchain.

+ [x] I read the contributing documentation.
+ [x] I am providing a screenshot of the (new feature) / (fixed bug).
<img width="1133" height="211" alt="image" src="https://github.com/user-attachments/assets/193d356d-cf49-4b03-a404-25d0e151484a" />
closes #3295

carlpinto25 committed:
improved readelf

carlpinto25 committed:
improved readelf

carlpinto25 committed:
replace readelf

carlpinto25 committed:
replace readelf

carlpinto25 committed:
fix errors in readelf.py

carlpinto25 committed:
fix: correct import order in test_readelf.py

carlpinto25 committed:
style: apply ruff formatting to test_readelf.py

carlpinto25 committed:
fix: remove unused imports

carlpinto25 committed:
fix: add type assertions for mypy in got.py

carlpinto25 committed:
feat: add symbol version support to readelf wrapper

disconnect3d commented on tests/unit_tests/test_readelf.py:
```diff
@@ -0,0 +1,49 @@
+
+# Mock gdb before importing pwndbg
+sys.modules["gdb"] = MagicMock()
+sys.modules["gdb"].VERSION = "12.1"
```
I believe this is wrong / not needed.

disconnect3d replied:
If we do need to mock gdb I think we have something done for it already somewhere?

carlpinto25 replied:
yes there is i updated the code

disconnect3d commented on tests/unit_tests/test_readelf.py:
```diff
@@ -0,0 +1,49 @@
+            assert isinstance(item["value"], int)
+            assert isinstance(item["name"], str)
+
+    print("test_get_got_entry passed")
```
This print is redundant

carlpinto25 replied:
my bad its removed now

disconnect3d commented on pwndbg/commands/got.py:
```diff
@@ -141,37 +141,29 @@ def _got(path: str, accept_readonly: bool, symbol_filter: str) -> None:
+            # Ensure types for mypy
+            assert isinstance(offset, int)
+            assert isinstance(value, int)
+            assert isinstance(name, str)
```
Can't we just do sth like `offset: int = ...` instead or it won't work?

carlpinto25 replied:
yes its possible i updated it

disconnect3d commented on pwndbg/wrappers/readelf.py:
```diff
@@ -17,18 +18,81 @@ class RelocationType(Enum):
-        for c in RelocationType:
-            if c.name in category:
-                entries[c].append(line)
+def get_got_entry(local_path: str) -> Dict[RelocationType, List[Dict[str, int | str]]]:
```
Why did you rewrite the logic here? Do the old logic with parsing lines had any bugs or something?

I get it we may want to return a structured output - that's fine - but - does the code do exactly the same stuff now?

disconnect3d commented on tests/unit_tests/test_readelf.py:
```diff
@@ -0,0 +1,49 @@
+            # Check types
+            assert isinstance(item["offset"], int)
+            assert isinstance(item["value"], int)
+            assert isinstance(item["name"], str)
```
Honestly, the test should really check for some values here instead of just checking if the types match

carlpinto25 replied:
Added checks for specific symbol names (puts, __libc_start_main) and GLIBC version presence

disconnect3d reviewed (changes requested):
Please see comments. I don't quite understand why did we change the `get_got_entry` parsing. I understand the want of it returning a dict, though a `NamedTuple` would be much better here, but I don't understand why did we change the parsing logic to something else.

Some questions and notes, apart from what is there in the comments:
1. Was there any deficiency or a bug with previous parsing logic?
2. The `tests/binaries/host/reference-binary` must be removed from the PR. Isn't it gitignored anyway?

carlpinto25 committed:
changes according to review

carlpinto25 committed:
test: add specific value assertions to test_readelf.py

- Added test_get_got_entry_direct() that tests pyelftools directly without complex mocks
- Checks for specific symbols: puts, __libc_start_main
- Verifies GLIBC version suffixes are present
- Fixed dbg mock to implement name() method

Addresses reviewer feedback to check actual values, not just types.

carlpinto25 committed:
style: fix import order in test_readelf.py

carlpinto25 committed:
Merge branch 'pwndbg:dev' into dev

carlpinto25 committed:
style: apply ruff formatting to test_readelf.py

carlpinto25 commented:
@disconnect3d The old logic had executed the system's readelf command, which is architecture-specific. When debugging an ARM binary on x86_64, the host's readelf fails to parse ARM-specific structures correctly, causing incorrect/missing GOT entries. This is literally issue #3295.

> Please see comments. I don't quite understand why did we change the `get_got_entry` parsing. I understand the want of it returning a dict, though a `NamedTuple` would be much better here, but I don't understand why did we change the parsing logic to something else.
>
Agreed it'd be cleaner. Should I change it in this PR?
> 2. The `tests/binaries/host/reference-binary` must be removed from the PR. Isn't it gitignored anyway?
>
i removed it

carlpinto25 committed:
removed unused import

carlpinto25 committed:
fixed error from mypy test

carlpinto25 committed:
fix: remove unused type: ignore comments

These are no longer needed as mypy no longer flags these pyelftools calls.

carlpinto25 committed:
revert: restore type: ignore comments needed for local mypy

The CI error about 'unused type: ignore' was incorrect - these comments
are required for mypy --strict to pass locally.

carlpinto25 commented:
@disconnect3d I'm seeing conflicting mypy requirements for the pyelftools calls in 
readelf.py
:

With # type: ignore[no-untyped-call]:
mypy --strict passes
mypy fails with "unused 'type: ignore' comment"

Without the comments:
mypy passes
mypy --strict fails with "Call to untyped function"

carlpinto25 committed:
Merge branch 'dev' into readelf

disconnect3d commented:
> @disconnect3d I'm seeing conflicting mypy requirements for the pyelftools calls in readelf.py :
> 
> With # type: ignore[no-untyped-call]: mypy --strict passes mypy fails with "unused 'type: ignore' comment"
> 
> Without the comments: mypy passes mypy --strict fails with "Call to untyped function"

Please make strict happy ideally by adding additional annotations instead of 'unused type ignore' comments.

Anyway, If one works and another fails, don't worry and let us know, we are going to fix this soon (we just noticed this problem too).

carlpinto25 committed:
temp fix for untyped calls error

carlpinto25 committed:
Merge branch 'dev' into readelf

carlpinto25 commented:
@disconnect3d  Thanks for the suggestion!
>Please make strict happy ideally by adding additional annotations instead of 'unused type ignore' comments.

adding annotations didnt work,snice nothing what i tried worked i did Mypy config override since it would be cleaner as temporary fix

carlpinto25 committed:
Merge branch 'dev' into readelf

disconnect3d committed:
Merge branch 'dev' into readelf

carlpinto25 committed:
Merge branch 'pwndbg:dev' into dev

disconnect3d commented on pyproject.toml:
```diff
@@ -251,6 +251,13 @@ module = [
+    "pwndbg.wrappers.readelf",
+    "pwndbg.commands.got",
+]
+disable_error_code = ["no-untyped-call"]
```
@k4lizen do we want those changes?

disconnect3d replied:
What do the tool mypy overrides module do even ugh

k4lizen replied:
we dont really want it but idk of a better solution so yeah its fine

jxuanli replied:
more annotations?

jxuanli replied:
or maybe add the `#[no-typed-call]` at the end of the effected lines rather the entire file

carlpinto25 committed:
Merge branch 'dev' into readelf

Resolved conflicts in:
- pwndbg/commands/got.py
- pwndbg/wrappers/readelf.py
- pyproject.toml
- tests/unit_tests/test_readelf.py

Kept symbol version support from readelf branch while integrating
ruff isort configuration and other updates from dev branch.

carlpinto25 committed:
Merge branch 'dev' into readelf

carlpinto25 committed:
Fix lint issues after merge

- Fix ruff formatting in test_readelf.py
- Add type assertions in got.py for mypy
- Regenerate documentation

carlpinto25 committed:
Fix CI lint errors

- Use modern type annotations (dict/list)
- Fix MockDebugger return type
- Remove unused type:ignore comments

carlpinto25 committed:
Add type:ignore for pyelftools in strict mode

CI runs mypy --strict which requires ignores for
untyped third-party library calls

carlpinto25 committed:
Disable warn_unused_ignores for tests

Fixes conflict between regular mypy and --strict:
- Strict mode requires type:ignore for untyped libs
- Regular mode warns about unused ignores
Solution: disable the warning for test files

jxuanli commented:
I'd also suggest that the readelf wrapper file should be removed since it is now technically not a wrapper anymore. It's one functions anyways and only gets invoked in one place so we should move it into commands/got.py

carlpinto25 committed:
Merge branch 'dev' into readelf

carlpinto25 committed:
Merge branch 'dev' into readelf

carlpinto25 commented:
@disconnect3d  all the tests pass now also what do you think of 

> I'd also suggest that the readelf wrapper file should be removed since it is now technically not a wrapper anymore. It's one functions anyways and only gets invoked in one place so we should move it into commands/got.py

disconnect3d commented:
> @disconnect3d all the tests pass now also what do you think of
> 
> > I'd also suggest that the readelf wrapper file should be removed since it is now technically not a wrapper anymore. It's one functions anyways and only gets invoked in one place so we should move it into commands/got.py

That sounds good / makes sense to me.

disconnect3d commented:
I will just merge this and open an issue for the moving of this to got.py

disconnect3d commented:
Created https://github.com/pwndbg/pwndbg/issues/3844
